### PR TITLE
Implemented times for SignalProducer

### PR DESCRIPTION
**Pull request by NachoSoto**
Completes part of #1751.

### CODE REVIEW

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 1]
```diff
@@ -552,10 +552,21 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
 public func switchMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalProducer<T>) -> SignalProducer<U>
 
-public func repeat<T>(count: Int)(producer: SignalProducer<T>) -> SignalProducer<T>
 public func retry<T>(count: Int)(producer: SignalProducer<T>) -> SignalProducer<T>
 */
 
+/// Repeats `producer` `count` times.
+/// Repeating `0` times results in a equivalent signal producer.
+public func repeat<T, E>(count: Int)(producer: SignalProducer<T, E>) -> SignalProducer<T, E> {
+	precondition(count >= 0)
```
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Alternatively I can do
```swift
switch count {
case 0:
	return producer
case let count where count > 0:
	return producer |> concat(producer |> repeat(count - 1))
default:
	fatalError("Invalid count: \(count)")
}
```

Let me know what you prefer :)

Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: A `precondition` like this is fine.

This actually raises an interesting question, though. My mental model of `count` was “the number of times you want to see the signal events,” but you've interpreted it as “the number of _additional_ times.”

The problem is that I don't think either interpretation is actually wrong. Maybe we should call this operator `times()` or something instead, so that it's clear which semantics it has?

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 2]
```diff
@@ -552,10 +552,21 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
+
+	if count == 0 {
+		return producer
+	} else {
+		return producer |> concat(producer |> repeat(count - 1))
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: This is clever, but for a huge count (e.g., `Int.max`) will assumably blow the stack, since `repeat()` recursively invokes itself immediately.

Can you refactor this to use iteration instead, and add a test that verifies that an overflow doesn't occur with an `Int.max` repeat?

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 3]
```diff
@@ -299,10 +299,52 @@ class SignalProducerSpec: QuickSpec {
 		}
 
 		describe("repeat") {
-			pending("should start a signal N times upon completion") {
+			it("should start a signal N times upon completion") {
+				let original = SignalProducer<Int, NoError>(values: [1, 2, 3])
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Can you please add spaces inside these array literals?

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 4]
```diff
@@ -299,10 +299,52 @@ class SignalProducerSpec: QuickSpec {
...
+			it("should not repeat upon error") {
+				let results: [Result<Int, TestError>] = [
+					success(1),
+					success(2),
+					failure(.Default)
+				]
+
+				var operationIndex = 0
+				let operation: () -> Result<Int, TestError> = {
+					assert(operationIndex < results.count, "Operation started too many times")
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Can this be an `expect()` instead? An assertion will probably bring down the test runner.

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 5]
```diff
@@ -299,10 +299,52 @@ class SignalProducerSpec: QuickSpec {
...
+				let expectedEvents: [Event<Int, TestError>] = [
+					.Next(Box(1)),
+					.Next(Box(2)),
+					.Error(Box(.Default))
+				]
+
+				// Can't test for equality because Array<T> is not Equatable,
+				// and neither is Event<T, E>.
+				expect(result.count).to(equal(expectedEvents.count))
+				expect(result[0] == expectedEvents[0]).to(beTruthy())
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Why can't this use `to(equal())`?

Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Could we just do this over the whole arrays instead?

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 6]
```diff
@@ -299,10 +299,52 @@ class SignalProducerSpec: QuickSpec {
...
+				let operation: () -> Result<Int, TestError> = {
+					assert(operationIndex < results.count, "Operation started too many times")
+
+					return results[operationIndex++]
+				}
+
+				let original = SignalProducer.try(operation)
+				let producer = original |> repeat(2)
+
+				let result = (producer |> materialize |> allValues).value!
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: We should avoid force-unwrapping in tests, because any failure will bring down the rest runner.

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift (hunk 5 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Notice my comment, it's not possible :/

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift (hunk 6 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Oh good call

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift (hunk 4 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Is there an equivalent to `except` that actually marks the test case as a failure and stops? If I just replace this with `except` the next line would be an out of bounds runtime crash.

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift (hunk 3 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: :+1:

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 1 above)
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: That's a good point, I think `times` is a good name for it :)

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift (hunk 5 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Never mind, I thought of a way to write this more concisely.

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift (hunk 4 above)
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Hrm, I don't know. I guess I would just change this into a conditional, then, and do `expect(false)` or something silly in the alternate case.

Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: :+1:

Comment (almassapargali) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: There is fail() afaik

**Comment by NachoSoto** [comment]
This is ready for review again :recycle:

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift [hunk 7]
```diff
@@ -298,14 +298,58 @@ class SignalProducerSpec: QuickSpec {
...
+		describe("times") {
+			it("should start a signal N times upon completion") {
+				let original = SignalProducer<Int, NoError>(values: [ 1, 2, 3 ])
+				let producer = original |> times(3)
+
+				let result = producer |> reduce([]) { $0 + [$1] } |> single
+				expect(result?.value).to(equal([ 1, 2, 3, 1, 2, 3, 1, 2, 3 ]))
+			}
+
+			it("should produce an equivalent signal producer if count is 0") {
```
Comment (kastiglione) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: "count is _1_"

**Comment by neilpa** [comment]
Continuing my comment from [here](https://github.com/ReactiveCocoa/ReactiveCocoa/issues/1762#issuecomment-74437536). Instead of `concat`, `times` should simply wrap the producer and restart it on completed events until `count` is exhausted.

File: ReactiveCocoaTests/Swift/SignalProducerSpec.swift (hunk 7 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpec.swift]: Good catch, changed!

**Comment by jspahrsummers** [comment]
I agree with @neilpa. Could you please make that change, @NachoSoto?

**Comment by NachoSoto** [comment]
I don't mind at all, but it's unfortunate that we need to implement it from scratch instead of relying on existing operators as an optimization. Or is there another reason I'm not seeing?

**Comment by jspahrsummers** [comment]
I don't think “not crashing from stack overflow” can be classified as merely an optimization. :stuck_out_tongue:

**Comment by NachoSoto** [comment]
I was thinking that it might be more appropriate to make concat not overflow, but now that I think about it, it's not common to concat many producers, but it might be very common to make a producer repeat infinitely, and that shouldn't be O(n).

Thanks guys! I'll make the change today :)

**Comment by neilpa** [comment]
I actually did this yesterday to make sure what I was suggesting worked - neilpa/ReactiveCocoa@81c2cf6d2da0e8c09dc77a082dd1a353b00a1ffc.  Go ahead and take that if you want.

**Comment by NachoSoto** [comment]
@neilpa haha, I had just finished writing it :) my implementation is slightly different, let me know if it's correct as well.

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 8]
```diff
@@ -552,10 +552,41 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
 
-public func repeat<T>(count: Int)(producer: SignalProducer<T>) -> SignalProducer<T>
 public func retry<T>(count: Int)(producer: SignalProducer<T>) -> SignalProducer<T>
 */
 
+/// Repeats `producer` a total of `count` times.
+/// Repeating `1` times results in a equivalent signal producer.
+public func times<T, E>(count: Int)(producer: SignalProducer<T, E>) -> SignalProducer<T, E> {
+	precondition(count >= 1)
+
```
Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Could short circuit on `1` and return `producer`

Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: True! changed

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 9]
```diff
@@ -552,10 +552,41 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
+			signal.observe(next: { value in
+				sendNext(observer, value)
+			}, error: { error in
+				sendError(observer, error)
+			}, completed: {
+				if count == 1 {
+					sendCompleted(observer)
+				} else {
+					let nextSignal = producer |> times(count - 1)
```
Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: `nextProducer` since this isn't a `Signal`

Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Actually you don't need this at all. There's a `startWithSignal` free function so you can chain it with `|>`.

**Comment by neilpa** [comment]
I'm not 100% on the `Disposable`s management but I think it's correct. I also like the way you handled the inner repeat over my approach of recursively calling the original `startWithSignal` closure.

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 9 above)
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Oh neat!

**Comment by NachoSoto** [comment]
Addressed @neilpa's comments. The implementation is much cleaner now :)

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 10]
```diff
@@ -552,10 +552,39 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
+
+		producer.startWithSignal { signal, signalDisposable in
+			serialDisposable.innerDisposable = signalDisposable
+
+			signal.observe(next: { value in
+				sendNext(observer, value)
+			}, error: { error in
+				sendError(observer, error)
+			}, completed: {
+				producer |> times(count - 1) |> startWithSignal { signal, signalDisposable in
```
Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Current style is to split these across multiple lines ([see single](https://github.com/NachoSoto/ReactiveCocoa/blob/signal-producer-repeat/ReactiveCocoa/Swift/SignalProducer.swift#L621-L637)) unless they're one-liners.

**Comment by neilpa** [comment]
One last note

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 11]
```diff
@@ -552,10 +552,41 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
 public func switchMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalProducer<T>) -> SignalProducer<U>
 
-public func repeat<T>(count: Int)(producer: SignalProducer<T>) -> SignalProducer<T>
 public func retry<T>(count: Int)(producer: SignalProducer<T>) -> SignalProducer<T>
 */
 
+/// Repeats `producer` a total of `count` times.
+/// Repeating `1` times results in a equivalent signal producer.
+public func times<T, E>(count: Int)(producer: SignalProducer<T, E>) -> SignalProducer<T, E> {
+	precondition(count >= 1)
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: I kinda think 0 should be equivalent to returning `empty`. Does that seem reasonable?

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 12]
```diff
@@ -552,10 +552,41 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
+				sendNext(observer, value)
+			}, error: { error in
+				sendError(observer, error)
+			}, completed: {
+				producer
+					|> times(count - 1)
+					|> startWithSignal { signal, signalDisposable in
+						serialDisposable.innerDisposable = signalDisposable
+						signal.observe(observer)
+				}
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Can you please fix the indentation here?

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 13]
```diff
@@ -552,10 +552,41 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
+		producer.startWithSignal { signal, signalDisposable in
+			serialDisposable.innerDisposable = signalDisposable
+
+			signal.observe(next: { value in
+				sendNext(observer, value)
+			}, error: { error in
+				sendError(observer, error)
+			}, completed: {
+				producer
+					|> times(count - 1)
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: It still doesn't seem like we need to be recursing here, especially because each recursion adds more memory footprint.

Can this be rewritten to use a `var` which tracks how many iterations are left, and simply invoke the outer `startWithSignal` again when there's more to do?

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 11 above)
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Yeah makes sense!

File: ReactiveCocoaTests/Swift/SignalProducerSpecHelpers.swift [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+//
+//  Created by Nacho Soto on 2/14/15.
+//  Copyright (c) 2015 GitHub. All rights reserved.
+//
+
+import ReactiveCocoa
+import LlamaKit
+import Nimble
+
+extension SignalProducer {
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpecHelpers.swift]: Can this please go into the actual spec file?

File: ReactiveCocoaTests/Swift/SignalProducerSpecHelpers.swift [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+	/// Creates a producer that can be started as many times as elements in `results`.
+	/// Each signal will immediately send either a value or an error.
+	static func tryWithResults<C: CollectionType where C.Generator.Element == Result<T, E>, C.Index.Distance == Int>(results: C) -> SignalProducer<T, E> {
```
Comment (jspahrsummers) [inline ReactiveCocoaTests/Swift/SignalProducerSpecHelpers.swift]: With the above change, can you please make this `private`?

**Comment by jspahrsummers** [comment]
Sorry, I found a few more things. :stuck_out_tongue_closed_eyes:

File: ReactiveCocoaTests/Swift/SignalProducerSpecHelpers.swift (hunk 14 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalProducerSpecHelpers.swift]: Absolutely! I wasn't sure what the best place for this was

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 13 above)
Comment (NachoSoto) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Done!

**Comment by NachoSoto** [comment]
I think I addressed all the comments :8ball:

File: ReactiveCocoa/Swift/SignalProducer.swift [hunk 16]
```diff
@@ -552,10 +552,51 @@ public func mergeMap<T, U>(transform: T -> SignalProducer<U>)(producer: SignalPr
...
+			producer.startWithSignal { signal, signalDisposable in
+				serialDisposable.innerDisposable = signalDisposable
+
+				signal.observe(next: { value in
+					sendNext(observer, value)
+				}, error: { error in
+					sendError(observer, error)
+				}, completed: {
+					if --remainingTimes > 0 {
+						iterate()
```
Comment (jspahrsummers) [inline ReactiveCocoa/Swift/SignalProducer.swift]: I think this will strongly capture the block within itself, creating a leak.

You might want @andymatuschak's [implementation of the Z combinator](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/63f55928b5cc613433bfe252cc66ee5ce69df8bd/ReactiveCocoa/Swift/SignalProducer.swift#L535-L539) instead.

**Comment by jspahrsummers** [comment]
:soon:

File: ReactiveCocoa/Swift/SignalProducer.swift (hunk 16 above)
Comment (neilpa) [inline ReactiveCocoa/Swift/SignalProducer.swift]: Or @robrix's [`fix`](https://github.com/robrix/Prelude/blob/master/Prelude/Fix.swift) which is what I had used previously. Either also has the added benefit of avoiding the two step initialization dance.

Comment (robrix) [inline ReactiveCocoa/Swift/SignalProducer.swift]: `Z` and `fix` are exactly equivalent, I think.

**Comment by NachoSoto** [comment]
Does [this](https://github.com/NachoSoto/ReactiveCocoa/commit/7e9063081082dbf8167167c02ce821ded758d446?w=0) look good?

**Comment by jspahrsummers** [comment]
Yup, that should do it! Just waiting on CI and then I'll merge this.

**Comment by NachoSoto** [comment]
Thank you!
Let me know if you think we should move the `fix` function somewhere else, happy to change that!

**Comment by NachoSoto** [comment]
:roller_coaster:? :D

**Comment by jspahrsummers** [comment]
Derp, sorry!
